fs-cache/src/cache.rs

```diff
@@ -0,0 +1,668 @@
+use std::fs;
+use std::num::NonZeroUsize;
+use std::path::{Path, PathBuf};
+
+use data_error::{ArklibError, Result};
+use fs_atomic_light::temp_and_move;
+use fs_storage::utils::extract_key_from_file_path;
+use lru::LruCache;
+
+/// A cache entry that stores a value and its size in bytes.
+///
+/// This structure is used to track both the actual data (value)
+/// and its memory usage (size) in the cache.
+struct CacheEntry<V> {
+    value: V,
+    size: usize,
+}
```

Comment:
We need to track memory consumption in bytes to be precise about when to offload values, and to support large values, too. We probably can't avoid saving value sizes into memory, otherwise when we hit the limit we cannot know how much values we need to offload.

However, that's where we could split the crate into 2 flavours:
1. dynamically-sized values e.g. byte vectors and text strings
2. statically-sized values e.g. integers

For the 2nd flavour we could utilize some standard Rust trait.

Is there a way to have these 2 flavours combined nicely in a single crate?

Reply:
> We need to track memory consumption in bytes to be precise about when to offload values...

Thtat's fine but we're actually reading the file size from the disk again in the `get_file_size` method, even though we already have it stored in the metadata. That seems a bit wasteful. Check out the next comment for more on this

> 1. dynamically-sized values e.g. byte vectors and text strings

That's a good point
I completely missed the dynamic types aspect when I looked at this. Now it makes a lot more sense why we need to track the data size instead of just the number of items.


> Is there a way to have these 2 flavours combined nicely in a single crate?

If we're dealing with types that have a fixed size, like `usize`, it doesn't really matter if we count how many items there are or just the total size they take up. But this completely breaks with the second flavour you mentioned.

The only solution I can think of right now is to treat all types of data as if they were the second type – basically, keep track of how much memory they use instead of how many there are. But that brings up the question of how to do this in a clean way

Main thread: https://github.com/ARK-Builders/ark-core/pull/95#discussion_r1864880931